web/srv/api_handlers.go

```diff
@@ -109,3 +116,69 @@ func (h *handler) handleApiStat(w http.ResponseWriter, req *http.Request, p http
 	}
 	renderJsonPb(w, result)
 }
+
+func (h *handler) handleApiTap(w http.ResponseWriter, req *http.Request, p httprouter.Params) {
+	ws, err := websocketUpgrader.Upgrade(w, req, nil)
+	if err != nil {
+		renderJsonError(w, err, http.StatusInternalServerError)
+		return
+	}
+	defer ws.Close()
+
+	var requestParams util.TapRequestParams
+	messageType, message, err := ws.ReadMessage()
+	if messageType == websocket.TextMessage {
+		err := json.Unmarshal(message, &requestParams)
+		if err != nil {
+			ws.WriteMessage(websocket.CloseMessage, []byte(err.Error()))
+			return
+		}
+	}
+
+	if requestParams.MaxRps == 0.0 {
+		requestParams.MaxRps = 1.0
+	}
+
+	tapReq, err := util.BuildTapByResourceRequest(requestParams)
+	if err != nil {
+		ws.WriteMessage(websocket.CloseMessage, []byte(err.Error()))
+		return
+	}
+
+	tapClient, err := h.apiClient.TapByResource(req.Context(), tapReq)
+	if err != nil {
+		ws.WriteMessage(websocket.CloseMessage, []byte(err.Error()))
+		return
+	}
+	defer tapClient.CloseSend()
+
+	go func() {
+		for {
+			rsp, err := tapClient.Recv()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				ws.WriteMessage(websocket.CloseMessage, []byte(err.Error()))
+				break
+			}
+
+			tapEvent := util.RenderTapEvent(rsp)
+			if err := ws.WriteMessage(websocket.TextMessage, []byte(tapEvent)); err != nil {
+				log.Error(err)
+				break
+			}
+		}
+	}()
+
+	for {
+		messageType, _, err := ws.ReadMessage()
+		if err != nil {
+			log.Error(err)
+			break
+		}
+		if messageType == websocket.CloseMessage {
+			break
+		}
+	}
```

Comment:
From looking at the gorilla documentation, it looks like all closes are returned as errors, even if they're expected. And we can check for expected closes using the [IsUnexpectedCloseError](https://godoc.org/github.com/gorilla/websocket#IsUnexpectedCloseError) (see example in the godoc). Based on that example, I'd write this as:

```
for {
	_, _, err := ws.ReadMessage()
	if err != nil {
		if websocket.IsUnexpectedCloseError(err, websocket.CloseNormalClosure) {
			log.Errorf("Unexpected close error: %s", err)
		}
		return
	}
}
```

Note that in order for this code to work correctly, you would need to start calling `ws.close(1000)` in the javascript code, to indicate a normal closure.